File: src/main/java/ru/translater/translate/yandex/YandexTranslationConfiguration.java
package ru.translater.translate.yandex;


import org.apache.http.HttpHeaders;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.http.converter.HttpMessageConverter;
import org.springframework.http.converter.json.MappingJackson2HttpMessageConverter;
import org.springframework.web.client.RestTemplate;

import java.util.ArrayList;
import java.util.List;

@Configuration
public class YandexTranslationConfiguration {
    private final String AUTHORIZATION_API_KEY = "Api-Key";

    @Bean
    public RestTemplate yandexRestTemplate(YandexConnectionProperties connectionProperties) {
        RestTemplate restTemplate = new RestTemplate();
        restTemplate.setMessageConverters(getMessageConverters());
        restTemplate.getInterceptors().add(
                (request, body, execution) -> {
                    request.getHeaders().add(HttpHeaders.AUTHORIZATION, AUTHORIZATION_API_KEY + " " + connectionProperties.getApiKey());
                    return execution.execute(request, body);
                }
        );

        return restTemplate;
    }

    private List<HttpMessageConverter<?>> getMessageConverters() {
        List<HttpMessageConverter<?>> converters =
                new ArrayList<>();
        converters.add(new MappingJackson2HttpMessageConverter());
        return converters;
    }

}

File: src/main/kotlin/com/translater/translate/yandex/configuration/YandexTranslationConfiguration.kt
package com.translater.translate.yandex.configuration

import com.translater.translate.yandex.properties.YandexPartConnectionProperties
import org.apache.http.HttpHeaders
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.http.HttpRequest
import org.springframework.http.client.ClientHttpRequestExecution
import org.springframework.http.client.ClientHttpRequestInterceptor
import org.springframework.http.converter.HttpMessageConverter
import org.springframework.http.converter.json.MappingJackson2HttpMessageConverter
import org.springframework.web.client.RestTemplate

@Configuration
class YandexTranslationConfiguration {
    companion object {
        private const val AUTHORIZATION_API_KEY = "Api-Key"
    }

    @Bean
    fun yandexRestTemplate(partConnectionProperties: YandexPartConnectionProperties): RestTemplate {
        return RestTemplate().apply {
            messageConverters = getCustomMessageConverters()
            interceptors.add(
                ClientHttpRequestInterceptor { request: HttpRequest, body: ByteArray, execution: ClientHttpRequestExecution ->
                    request.headers.add(
                        HttpHeaders.AUTHORIZATION,
                        "$AUTHORIZATION_API_KEY ${partConnectionProperties.apiKey}"
                    )
                    execution.execute(request, body)
                }
            )
        }
    }

    private fun getCustomMessageConverters(): List<HttpMessageConverter<*>> {
        val converters: MutableList<HttpMessageConverter<*>> = mutableListOf()
        converters.add(MappingJackson2HttpMessageConverter())
        return converters
    }
}